ads/google/a4a/utils.js

```diff
@@ -0,0 +1,317 @@
+ * @return {!Uint8Array}
+ * @visibleForTesting
+ */
+export function base64ToByteArray(str) {
+  const bytesAsString = atob(str);
+  const len = bytesAsString.length;
+  const bytes = new Uint8Array(len);
+  for (let i = 0; i < len; i++) {
+    bytes[i] = bytesAsString.charCodeAt(i);
+  }
+  return bytes;
+}
+
+/**
+ * @param {!ArrayBuffer} creative
+ * @param {!Headers} responseHeaders
+ * @return {!Promise<!AdResponseDef>}
+ */
+export function extractGoogleAdCreativeAndSignature(
+    creative, responseHeaders) {
+  let signature = null;
+  try {
+    if (responseHeaders.has(AMP_SIGNATURE_HEADER)) {
+      signature = base64ToByteArray(responseHeaders.get(AMP_SIGNATURE_HEADER));
+    }
+  } finally {
+    return Promise.resolve({creative, signature});
+  }
+}
+
+/**
+ * @param {!AmpA4A} a4a
+ * @param {string} baseUrl
+ * @param {number} startTime
+ * @param {number} slotNumber
+ * @param {!Array<!QueryParameter>} queryParams
+ * @param {!Array<!QueryParameter>} unboundedQueryParams
+ * @param {(string|undefined)} clientId
+ * @param {string} referrer
+ * @return {string}
+ */
+function buildAdUrl(
+    a4a, baseUrl, startTime, slotNumber, queryParams, unboundedQueryParams,
+    clientId, referrer) {
+  const global = a4a.getWin();
+  const documentInfo = documentInfoFor(global);
+  if (!global.gaGlobal) {
+    // Read by GPT for GA/GPT integration.
+    global.gaGlobal = {
+      vid: clientId,
+      hid: documentInfo.pageViewId,
+    };
+  }
+  const viewport = viewportFor(global);
+  const slotRect = viewport.getLayoutRect(a4a.element);
+  const viewportRect = viewport.getRect();
+  const iframeDepth = iframeNestingDepth(global);
+  const dtdParam = {name: 'dtd'};
+  const allQueryParams = [
+    ...queryParams,
+    {
+      name: 'is_amp',
+      value: a4a.supportsShadowDom() ?
+          AmpAdImplementation.AMP_AD_XHR_TO_IFRAME_OR_AMP :
+          AmpAdImplementation.AMP_AD_XHR_TO_IFRAME,
+    },
+    {name: 'amp_v', value: '$internalRuntimeVersion$'},
+    {name: 'dt', value: startTime},
+    {name: 'adk', value: adKey(slotNumber, slotRect, viewportRect)},
+    {name: 'c', value: makeCorrelator(clientId, documentInfo.pageViewId)},
+    {name: 'output', value: 'html'},
+    {name: 'nhd', value: iframeDepth},
+    {name: 'eid', value: a4a.element.getAttribute('data-experiment-id')},
+    {name: 'bih', value: viewportRect.height},
+    {name: 'biw', value: viewportRect.width},
+    {name: 'adx', value: slotRect.left},
+    {name: 'ady', value: slotRect.top},
+    {name: 'u_hist', value: getHistoryLength(global)},
+    dtdParam,
+    ...unboundedQueryParams,
+    {name: 'url', value: documentInfo.canonicalUrl},
+    {name: 'top', value: iframeDepth ? topWindowUrlOrDomain(global) : null},
+    {
+      name: 'loc',
+      value: global.location.href == documentInfo.canonicalUrl ?
+          null : global.location.href,
+    },
+    {name: 'ref', value: referrer},
+  ];
+  dtdParam.value = elapsedTimeWithCeiling(timer.now(), startTime);
```

Comment:
Indeed it is. But it also needs to happen as late in the process as possible so that we can get accurate timing. Do you have any suggestions for how to improve it?